### PR TITLE
CDAP-4075 Storing node states of all nodes in the run record of the Workflow.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-4075

This PR only adds the implementation of the storing the state of all nodes inside the RunRecord of the Workflow. The current implementation is for MapReduce program. Spark will be done in next PR.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 1]
```diff
@@ -317,7 +318,8 @@ private void recordCompletedWorkflow(final Id.Workflow id, String pid) {
     Map<String, WorkflowNode> nodeIdMap = workflowSpec.getNodeIdMap();
     final List<WorkflowDataset.ProgramRun> programRunsList = new ArrayList<>();
     for (Map.Entry<String, String> entry : run.getProperties().entrySet()) {
-      if (!("workflowToken".equals(entry.getKey()) || "runtimeArgs".equals(entry.getKey()))) {
+      if (!("workflowToken".equals(entry.getKey()) || "runtimeArgs".equals(entry.getKey())
+        || "workflowNodeState".equals(entry.getKey()))) {
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java]: This logic needs to be refactored. Filed a JIRA - https://issues.cask.co/browse/CDAP-5319

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT3723-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 2]
```diff
@@ -354,6 +355,18 @@ void setWorkflowProgramStart(Id.Program programId, String programRunId, String w
...
 
   /**
+   * Logs the stop of the program running under Workflow.
+   * @param workflowRunId workflow run which started the program
+   * @param programRunId program run started by the Workflow
+   * @param stopTimeInSeconds stop time in seconds
+   * @param runStatus final status of the program
+   * @param failureCause cause of failure if the node failed, null otherwise
+   */
+
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: nit: Remove newline

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Removed.

**Comment by sagarkapare** [comment]
Build is green. @chtyim Please take a look. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java [hunk 3]
```diff
@@ -165,6 +165,11 @@ private ApplicationSpecificationAdapter(SchemaGenerator schemaGenerator, Gson gs
         SortedMap.class.isAssignableFrom(rawType)) {
         return null;
       }
+
+      if (!ParameterizedType.class.isAssignableFrom(type.getType().getClass())) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java]: What is this for?

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  public WorkflowNodeState(String nodeId, NodeStatus nodeStatus, @Nullable String runId,
+                           @Nullable String failureCause) {
+    this.nodeId = nodeId;
+    this.nodeStatus = nodeStatus;
+    this.runId = runId;
+    this.failureCause = failureCause;
+  }
+
+  public String getNodeId() {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java]: Public methods in public API need javadoc.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 5]
```diff
@@ -354,6 +355,17 @@ void setWorkflowProgramStart(Id.Program programId, String programRunId, String w
...
   /**
+   * Logs the stop of the program running under Workflow.
+   * @param workflowRunId workflow run which started the program
+   * @param programRunId program run started by the Workflow
+   * @param stopTimeInSeconds stop time in seconds
+   * @param runStatus final status of the program
+   * @param failureCause cause of failure if the node failed, null otherwise
+   */
+  void setWorkflowProgramStop(ProgramRunId workflowRunId, ProgramRunId programRunId, long stopTimeInSeconds,
+                              ProgramRunStatus runStatus, @Nullable String failureCause);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: What does the failureCause String carries? The message in the `Throwable`? Or something else? The `String` type seems pretty unclear on what is actually there.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 6]
```diff
@@ -231,35 +240,46 @@ public void starting() {
...
       @Override
       public void terminated(Service.State from) {
         closeAllQuietly(closeables);
+        ProgramRunStatus runStatus = ProgramController.State.COMPLETED.getRunStatus();
         if (from == Service.State.STOPPING) {
           // Service was killed
+          runStatus = ProgramController.State.KILLED.getRunStatus();
+        }
+
+        if (workflowRunId == null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: This logic is pretty brutal. Is there a way for the Store.setStop to figure out if a given `ProgramRunId` has a workflow as parent and just do the related RunRecords update as well? Ideally the same should apply for the setStart as well.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 7]
```diff
@@ -231,35 +240,46 @@ public void starting() {
...
+                                       TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()), runStatus, null);
         }
       }
 
       @Override
       public void failed(Service.State from, Throwable failure) {
         closeAllQuietly(closeables);
-        store.setStop(program.getId(), runId.getId(), TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                      ProgramController.State.ERROR.getRunStatus());
+        if (workflowRunId == null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: This is almost identical to terminate. Maybe better to have a private method in the listener to do so.

File: cdap-proto/src/main/java/co/cask/cdap/proto/ProgramRunStatus.java [hunk 8]
```diff
@@ -26,5 +28,26 @@
...
-  KILLED
+  KILLED;
+
+  /**
+   * Conversion from program run status to Workflow node status.
+   * @param status the program run status to be converted
+   * @return the converted Workflow node status
+   */
+  public static WorkflowNodeState.NodeStatus toNodeStatus(ProgramRunStatus status) {
+    switch(status) {
```
Comment (chtyim) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ProgramRunStatus.java]: Seems like `NodeStatus` is just a subset of `ProgramRunStatus`, is it? If that's the case, can we just use `ProgramRunStatus`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java]: When I want to update the run record with the WorkflowNodeState, the corresponding code is 
```java
    Map<String, String> propertiesToUpdate = new HashMap<>();
    propertiesToUpdate.putAll(runRecordMeta.getProperties());
    propertiesToUpdate.put(WORKFLOW_NODE_STATE_KEY, GSON.toJson(nodeStates));
```
Conversion of `nodeStates` to json using GSON instance fails. This happens because when the `ApplicationSpecificationAdapter.create` is called type parameter passed to it is of type `java.util.HashMap` casting of which to `ParameterizedType` fails. 

Another approach to fix it is use `new Gson()` instead of static `GSON`.

File: cdap-proto/src/main/java/co/cask/cdap/proto/ProgramRunStatus.java (hunk 8 above)
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ProgramRunStatus.java]: Since we want the node status map to be returned from the `WorkflowContext`, we want this enum in cdap-api which cannot have classes from cdap-proto.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java (hunk 5 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Changed API to accept the `Throwable` now, however while persisting it in the RunRecord using the detail message from the exception.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 6 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Since systemArgs contain the Workflow related details, using them in the store to update the Workflow run accordingly. So that we only need the generic `setStart` and `setStop` methods.

**Comment by sagarkapare** [comment]
New build is running here - http://builds.cask.co/browse/CDAP-DUT3723-4. @chtyim Please take a look again. Thanks!

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  private final NodeStatus nodeStatus;
+  private final String runId;
+  private final String failureCause;
+
+  /**
+   * Create a new instance.
+   * @param nodeId id of the node inside the Workflow
+   * @param nodeStatus status of the node
+   * @param runId run id assigned to the node, null if current node represents custom action or predicate
+   * @param failureCause cause of failure, null if execution of the node succeeded
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java]: So what does it carry? Is it just the error message? Why can't we provide the actual cause to the user since this object is part of the programatic API that the user can use to retrieve node states through the WorkflowContext, right? 

I understand that currently we use the same class to serialize the information and store it in the MDS. However, it may not be the right thing to do. Also, there is a way to serialize and deserialize a Throwable (for information purpose, not really reconstructing the real stacktrace, since it's not possible). See `LogThrowable` and `LogThrowableCodec` in twill.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java (hunk 3 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java]: I don't understand what you mean. Why it is just a HasMap without the type parameters? You can use `GSON.toJson(nodeStates, new TypeToken<Map<String, NodeState>>() { }.getType())`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 10]
```diff
@@ -231,35 +229,28 @@ public void starting() {
...
-                        ProgramController.State.COMPLETED.getRunStatus());
+          runStatus = ProgramController.State.KILLED.getRunStatus();
         }
+
+        store.setStop(program.getId(), runId.getId(), TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
+                      runStatus, null);
       }
 
       @Override
       public void failed(Service.State from, Throwable failure) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: `@Nullable` on `failure`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 11]
```diff
@@ -166,14 +166,14 @@ public void init(ProgramController.State state, @Nullable Throwable cause) {
         public void completed() {
           LOG.debug("Program {} {} {} completed successfully.", id.getNamespaceId(), id.getApplicationId(), id.getId());
           store.setStop(id, runId, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                        ProgramController.State.COMPLETED.getRunStatus());
+                        ProgramController.State.COMPLETED.getRunStatus(), null);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Better to have another `store.setStop` method that doesn't need the failure cause to pass it and default it to `null`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 10 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Added.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java (hunk 11 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Added another version of the setStop method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java]: Fixed.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java (hunk 9 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java]: Will take a look at it and do as a separate PR.

**Comment by sagarkapare** [comment]
New build is running here - http://builds.cask.co/browse/CDAP-DUT3723-7

**Comment by chtyim** [comment]
LGTM

**Comment by sagarkapare** [comment]
Build is green. Merging.
